File: hm10_client.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "btlib.h" 

int notify_callback(int node,int cticn,unsigned char *dat,int datlen);
void getstring(char *prompt,char *s,int len);
int inputint(char *ps);

int main()
  {
  int node,cticn;
  unsigned char uuid[2];
  char *buf;
  
  if(init_blue("devices.txt") == 0)
    return(0);
    
  printf("\nScanning for HM10\n");
  le_scan();
  printf("\nEnter [ ] to scroll device list\n");
  node = inputint("Enter node of HM10 (should report UUID=FFE0)");
  if(node < 0)
    return(0);
  if(connect_node(node,CHANNEL_LE,0) == 0)
    {
    printf("Connect failed\n");
    return(0);
    }
    // read services
  find_ctics(node);
    // find data characteristic UUID=FFE1
  uuid[0] = 0xFF;
  uuid[1] = 0xE1;
  cticn = find_ctic_index(node,UUID_2,uuid);
  if(cticn > 0)
    {
    printf("Found data characteristic index = %d\n",cticn);
    printf("Enabling notifications\n");
    notify_ctic(node,cticn,NOTIFY_ENABLE,notify_callback);
    printf("Sending Hello\n");
    buf = "Hello\n";
    write_ctic(node,cticn,buf,strlen(buf));
    printf("Waiting 30s for a reply from HM10\n");
    printf("Send an ASCII string reply from the HM10\n");
    read_notify(30000);
    printf("\n30s wait for reply timed out\n");
    }
  else
    printf("Data characteristic FFE1 not found\n");
   
  printf("Disconnecting HM10\n");
  disconnect_node(node);
  close_all();
  return(0);
  }  
  
int notify_callback(int node,int cticn,unsigned char *dat,int datlen)
  {
  int n;

  // print hex
  for(n = 0 ; n < datlen ; ++n)
    printf(" %02X",dat[n]);
  printf("\n");
  // print ASCII - dat may not be zero terminated 
  for(n = 0 ; n < datlen ; ++n)
    {
    if(dat[n] >= 32 || dat[n] == 10 || dat[n] == 13)
      printf("%c",dat[n]);
    else
      printf(".");
    }
  printf("\n");
  }  

void getstring(char *prompt,char *s,int len)
  {
  int n;
  
  do
    {
    printf("%s",prompt);
    fgets(s,len,stdin);
    }
  while(s[0] == 10);

  n = 0;
  while(s[n] != 10 && s[n] != 13 && s[n] != 0)
    ++n;
  s[n] = 0;
  }  
     
/***** INPUT INTEGER ********/     
     
int inputint(char *ps)
  {
  int n,flag;
  char s[128];
 
  do
    {
    printf("%s  (x=cancel)\n",ps);
  
    getstring("? ",s,128);
       
    flag = 0;
    for(n = 0 ; s[n] != 0 ; ++n)
      {
      if(s[n] < '0' || s[n] > '9')
        flag = 1;
      }
    if(flag == 0)
      n = atoi(s);
    else if(s[0] == 'x')
      {
      n = -1;
      flag = 0;
      }
    else if(s[0] == '[')
      scroll_back();
    else if(s[0] == ']')
      scroll_forward();
    else
      printf("Not a number\n");
    }
  while(flag != 0);
  return(n);
  } 

File: hm10_client.py
import btfpy


def notify_callback(node,cticn,dat,datlen):
  print("Reply from HM10")
  btfpy.Print_data(dat)
  # end notify_callback

def inputint(ps):
  flag = 1
  val = -1
  while flag != 0:
    print(ps + "  (x=cancel)")
    s = input("? ")
    flag = 0
    if(len(s) > 0):
      for n in range(len(s)):
        if(s[n] < '0' or s[n] > '9'):
          flag = 1
      if flag == 0:
        val = int(s)
      elif s[0] == 'x':
        val = -1
        flag = 0
      elif s[0] == '[':
        btfpy.Scroll_back()
      elif s[0] == ']':
        btfpy.Scroll_forward()
      else:
        print("Not a number")
    else:
      flag = 1
    # end while flag != 0  
 
  return(val)
  # end inputint


##### START #####
  
if(btfpy.Init_blue("devices.txt") == 0):
  exit(0)
    
print("Scanning for HM10")
btfpy.Le_scan()

print("Enter [ ] to scroll device list")
node = inputint("Enter node of HM10 (should report UUID=FFE0)")
if(node < 0):
  exit(0)

if(btfpy.Connect_node(node,btfpy.CHANNEL_LE,0) == 0):
  print("Connect failed")
  exit(0)
  
  # read services
btfpy.Find_ctics(node)
  # find data characteristic UUID=FFE1
cticn = btfpy.Find_ctic_index(node,btfpy.UUID_2,[0xFF,0xE1])
if(cticn > 0):
  print("Found data characteristic FFE1 index")  
  print("Enabling notifications")
  btfpy.Notify_ctic(node,cticn,btfpy.NOTIFY_ENABLE,notify_callback)
  print("Sending Hello")
  btfpy.Write_ctic(node,cticn,"Hello\n",0)
  print("Waiting 30s for a reply from HM10")
  print("Send an ASCII string reply from the HM10")
  btfpy.Read_notify(30000)
  print("30s wait for reply timed out")
else:
  print("Data characteristic FFE1 not found")
   
print("Disconnecting HM10")
btfpy.Disconnect_node(node)
btfpy.Close_all()

